capsules/src/net/udp/driver.rs

```diff
@@ -0,0 +1,473 @@
+    ///
+    /// ### `subscribe_num`
+    ///
+    /// - `0`: Setup callback for when frame is received.
+    /// - `1`: Setup callback for when frame is transmitted.
+    fn subscribe(
+        &self,
+        subscribe_num: usize,
+        callback: Option<Callback>,
+        app_id: AppId,
+    ) -> ReturnCode {
+        match subscribe_num {
+            0 => self.do_with_app(app_id, |app| {
+                app.rx_callback = callback;
+                ReturnCode::SUCCESS
+            }),
+            1 => self.do_with_app(app_id, |app| {
+                app.tx_callback = callback;
+                ReturnCode::SUCCESS
+            }),
+            _ => ReturnCode::ENOSUPPORT,
+        }
+    }
+
+    /// UDP control
+    ///
+    /// ### `command_num`
+    ///
+    /// - `0`: Driver check.
+    /// - `1`: Get the interface list
+    ///        app_cfg (out): 16 * `n` bytes: the list of interface IPv6 addresses, length
+    ///                       limited by `app_cfg` length.
+    /// - `2`: Transmit payload
+    fn command(&self, command_num: usize, arg1: usize, _: usize, appid: AppId) -> ReturnCode {
+        match command_num {
+            0 => ReturnCode::SUCCESS,
+
+            //  Writes the requested number of network interface addresses
+            // `arg1`: number of interfaces requested that will fit into the buffer
+            1 => self.do_with_cfg_mut(appid, arg1 * mem::size_of::<IPAddr>(), |cfg| {
+                let n_ifaces_to_copy = cmp::min(arg1, INTERFACES.len());
+                let iface_size = mem::size_of::<IPAddr>();
+                for i in 0..n_ifaces_to_copy {
+                    cfg[i * iface_size .. (i+1) * iface_size].copy_from_slice(&INTERFACES[i].0);
+                }
+                // Returns total number of interfaces
+                ReturnCode::SuccessWithValue { value: INTERFACES.len() }
+            }),
+
+            // Transmits UDP packet stored in
+            2 => {
+                self.do_with_app(appid, |app| {
+                    if app.pending_tx.is_some() {
+                        // Cannot support more than one pending tx per process.
+                        return ReturnCode::EBUSY;
+                    }
+                    let next_tx = app.app_cfg.as_ref().and_then(|cfg| {
+                        if cfg.len() != 2 * mem::size_of::<IPAddrPort>() {
+                            // debug!("cfg len is {:?}, needed at least {:?}", cfg.len(), 2 * mem::size_of::<IPAddrPort>());
+                            return None;
+                        }
+
+                        debug!("{:?}", self.parse_ip_port_pair(&cfg.as_ref()[mem::size_of::<IPAddrPort>()..]));
+
+                        if let (Some(dst), Some(src)) = (self.parse_ip_port_pair(&cfg.as_ref()[mem::size_of::<IPAddrPort>()..]),
+                                                         self.parse_ip_port_pair(&cfg.as_ref()[..mem::size_of::<IPAddrPort>()])) {
+                            Some([src, dst])
```

Comment:
What does the app pass as the src address? Is it used anywhere?

Reply:
Right now it is not used anywhere -- it is eventually replaced by the src IP stored in the kernel. I believe at the time this was designed it was unclear whether we would let the src IP addr be chosen by userland apps. I will remove the src address field and add documentation that the
src IP is set in the kernel when I consolidate all of the configuration of the network stack.